### PR TITLE
Use Gradle to entirely manage project dependency #5508

### LINKED ISSUE #5508: Use Gradle to entirely manage project dependency
The time's finally come to get rid of the zip library for good.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #5508 

It's here! The advantages brought by this is immense.

- Zip library is no longer needed. Gradle will manage all dependencies, and a task has been written to copy all necessary ones to the output directory. This will make adding and updating dependencies relatively effortless.
- Oh, and do I mention that the jar sources are automatically attached to the jars?
- Removed so many declared libraries which are unused. The transitive dependencies of the declared libraries are sufficient to cover everything we need locally. In staging server, the uploading of profile pictures and admin email attachments as well as accessing Gmail inbox have been tested.
- Building and running the app via Gradle is now supported, simply with `./gradlew appengineRun` or `gradlew.bat appengineRun` command. We're almost completely independent of Eclipse now (not saying we can ditch it; some tasks are far better done with Eclipse).

Now on for some technicalities:
- `.classpath` is now git-ignored. The irony is, it's the very first entry of our `.gitignore` file all these while. This is easily regenerated with `./gradlew eclipseClasspath` or `gradlew.bat eclipseClasspath` command, and is configured with the compile-time dependencies from both main and test.
- The `.template.prefs` is no longer applicable because the libraries now come from the `.gradle` directory, which is even harder to predict. We just have to live with Eclipse warnings due to Google Web App problem, which is not a big price to pay.
- The copy-and-rename and remove duplicate libs are there because GAE-Eclipse imposed certain names for their library files inside the output folder.
- The resolution strategy for `common-codec` is unnecessary. For GAE SDK, it's needed because apparently one of the libs have dependency to latest GAE SDK.
- The fixes for some Java files are needed because of a crime we've committed: production code using non-production library. They'd work perfectly fine locally, but on production they'll fail gracefully. The previous`build.gradle` config sneakily masked this crime (fixed now), while Eclipse can't even differentiate them.

~~Left to do: write a doc on managing dependencies.~~ A documentation on managing dependencies is also provided.

### CODE REVIEW

**Comment by wkurniawan07** [comment]
I have tested the processes on both my Windows 8.1 and Mac OS X 10.11.2, both via Eclipse and via Gradle (to simulate Travis environment), and everything works out fine on my side, but I need more people to test this new routine.

@kanghj @LowWeiLin @ashrayjain @JosephineKwa @YongJieYongJie @chowyb will appreciate your help! Here's how you can help:

1. Checkout to this branch.
2. Delete all the current library files, in `src/main/webapp/WEB-INF/lib` and `src/test/resources/lib`.
3. Run the command `./gradlew setup` or `gradlew.bat setup`, whichever one applies.
4. Refresh the project in Eclipse, run the dev server, and do a full test run.
5. (Optional) Try building and testing via Gradle commands as well, the commands are `appengineRun` and `travisTests` respectively.

If there are problems (there shouldn't be), I'll try to help as best as I can.

**Comment by chowyb** [comment]
@wkurniawan07 it works for me! :)

**Comment by JosephineKwa** [comment]
Works fine for me!

**Comment by damithc** [comment]
Can still deploy from Eclipse?

**Comment by wkurniawan07** [comment]
@damithc our Eclipse workflow is as intact as it can be.

**Comment by kanghj** [comment]
@wkurniawan07, works for me on ubuntu too, but didn't work right away. I have java 8 as the java version in my path, and after running `./gradlew setup`, eclipse ran teammates on java 8 which caused problems. Might need to add an additional step to configure eclipse to use java 7 if the person setting up is using Java 8 as the default java version?

**Comment by wkurniawan07** [comment]
@kanghj just to ensure we're on the same page, this is because of `.classpath` right? That should be correct-able via Eclipse.
Also, can you try this: get `gradle.properties` from the template, specify JDK 7 in the `org.gradle.java.home` variable, and regenerate the `.classpath` file again. Not sure if it'll help, but worth a shot.

**Comment by kanghj** [comment]
Yes, it's because of .classpath. Yup, I got it to work by configuring it in eclipse. Specifying `org.gradle.java.home` works too.

**Comment by wkurniawan07** [comment]
I knew at some point we have to make use of this `gradle.properties` :p @kanghj everything else working, e.g Eclipse can build/run tests/etc. without problem?

**Comment by kanghj** [comment]
Yup, managed to build and pass all tests without problem. :+1:

**Comment by YongJieYongJie** [comment]
A bit late, but works on my Lubuntu too (after increasing its partition size from 8gb to 12gb)

**Comment by wkurniawan07** [comment]
I think this can call for a proper review already.

File: devdocs/dependencies.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (damithc) [inline devdocs/dependencies.md]: Does this page linked from other relevant pages?

File: devdocs/dependencies.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+There are two factors to consider:
+- Whether the library is for production or non-production code
+- Whether the library is needed for compile-time or runtime only
+
+Add the library in the appropriate section in the `build.gradle` file.
+
+## Updating libraries
+
+> Note the following:
+- Change `./gradlew` to `gradlew.bat` in Windows.
```
Comment (damithc) [inline devdocs/dependencies.md]: I thought Windows runs `gradlew.bat` even if you type only `gradlew`? I always omit the `.bat` part and it still works.

File: devdocs/dependencies.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+The Eclipse `.classpath` needs to be updated whenever there is a change to compile-time library.
+Here are the steps:
+
+1. Delete the existing `.classpath` file.
+2. Run the command `./gradlew eclipseClasspath`.
+3. If there is an Eclipse instance running the project, you need to refresh it for the changes to take effect.
+
+### Updating output directory, `WEB-INF/lib`
+
+This directory needs to be updated whenever there is a change to production library.
```
Comment (damithc) [inline devdocs/dependencies.md]: `to production libraries`

File: gradle.template.properties [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# Use this property to enable/disable Gradle Daemon (long-lived background process)
+# Reference: https://docs.gradle.org/current/userguide/gradle_daemon.html
+org.gradle.daemon=false
+
+# Use this property if you want to use a JDK other than the one specified in your PATH variable.
+# org.gradle.java.home=
```
Comment (damithc) [inline gradle.template.properties]: Give a sample value to this property?

**Comment by damithc** [comment]
Added some comments. 
Also, it is not clear what regular devs need to do when there is a change to dependencies. Do they have to do anything at all? How are they notified?

File: devdocs/dependencies.md (hunk 2 above)
Comment (wkurniawan07) [inline devdocs/dependencies.md]: I'm guessing older Windows version didn't support that. http://superuser.com/questions/319430/run-a-bat-file-without-typing-the-extension

File: devdocs/dependencies.md (hunk 1 above)
Comment (wkurniawan07) [inline devdocs/dependencies.md]: Hmmm what pages would be relevant to be linked?

**Comment by wkurniawan07** [comment]
@damithc I'm afraid I don't solve that problem; it will still be as before, either the dev realizes it him/herself when the missing lib causes compile error, or announce via some channel. I also don't foresee any easy way to solve it with (or without) any build tool.

It should be noted that this is an overhead caused by our Eclipse workflow. If we depend entirely on Gradle for building and testing, all dependencies will resolve automatically (like how we can see on Travis). Again, I'm not trying to say that our Eclipse workflow needs to be rid of, but to live with both we have to deal with certain overheads like this.

**Comment by wkurniawan07** [comment]
Changes made:
- Example values for `org.gradle.java.home`
- Added a task to force-reset dependencies (acting on both the `.classpath` and `WEB-INF/lib` directory). Safe to be run multiple times. It will simply grab the jar copies from local Gradle cache so there are no redundant downloads.

**Comment by damithc** [comment]
The doc can be linked from README (under 'supplementary documentation'?). It's nice if all our docs are linked from README so that it acts like a site map.

**Pull request merged**